Add optional zerocopy support

zebreus opened the pull request:
This PR adds a flag that enables implementations for [zerocopy](https://github.com/google/zerocopy) traits on BitFlags.

Converting a `BitFlags` into the underlying number is trivial and always safe, so we can just derive those traits. Converting from a number to `BitFlags` is more tricky, because not every bit-pattern is valid. To assert that the bit-pattern is valid this PR provides a custom implementation of the `zerocopy::TryFromBytes` trait. Custom implementations for zerocopy traits `TryFromBytes` seem [somewhat discouraged but supported](https://github.com/google/zerocopy/issues/287), so it should be fine to implement that.

zebreus committed:
Fix formatting

zebreus committed:
Add zerocopy support

meithecatte commented on src/lib.rs:
```diff
@@ -1032,3 +1042,69 @@ mod impl_serde {
+        T::Numeric: Immutable,
+        T::Numeric: FromZeros,
+    {
+        fn only_derive_is_allowed_to_implement_this_trait() {}
```
I think this could use a comment linking to google/zerocopy#287, so that this doesn't scream "crimes! damn crimes!" at anyone who looks at this later.

meithecatte reviewed (changes requested):
For now I'm not looking at the details of how the zerocopy API works. I'll take a closer look later.

zebreus committed:
Fix zerocopy validation logic

zebreus committed:
Add a comment to justify implementing scary things

With the zerocopy feature activated we implement some zerocopy
traits containing an `only_derive_is_allowed_to_implement_this_trait`
function. That is fine as long as we know what we are doing.

meithecatte commented on src/lib.rs:
````diff
@@ -667,6 +667,33 @@ where
+    /// assert_eq!(BitFlags::<MyFlag>::validate_bits(0b0100), false);
+    /// assert_eq!(BitFlags::<MyFlag>::validate_bits(0b1111), false);
+    /// ```
+    #[must_use]
+    #[inline(always)]
+    pub fn validate_bits(bits: T::Numeric) -> bool {
+        // SAFETY: We're truncating out all the invalid bits so it will
+        // only be different if there are invalid bits set.
+        (bits & T::ALL_BITS) == bits
+    }
````
Honestly, I'm not sure if this makes sense as a public API.

meithecatte commented on src/lib.rs:
```diff
@@ -1102,8 +1129,11 @@ mod impl_zerocopy {
+                // Dereference the pointer to the candidate
+                let candidate =
+                    my_candidate.read_unaligned::<zerocopy::pointer::BecauseImmutable>();
+                return BitFlags::<T>::validate_bits(candidate);
```
```suggestion
                BitFlags::<T>::validate_bits(candidate)
```

meithecatte commented on src/lib.rs:
```diff
@@ -1102,8 +1129,11 @@ mod impl_zerocopy {
             {
-                (my_candidate.read_unaligned::<zerocopy::pointer::BecauseImmutable>() ^ T::ALL_BITS)
-                    == T::EMPTY
+                // TODO: Currently this assumes that the candidate is aligned. We actually need to check this beforehand
```
Not sure if this is something I would've caught. Do you have any links I could reference for example zerocopy implementations for similar types?

joshlf commented on src/lib.rs:
```diff
@@ -1047,3 +1069,84 @@ mod impl_serde {
+        }
+
+        #[inline]
+        fn is_bit_valid<
```
Looks like this impl relies on semver-unstable `#[doc(hidden)]` internals, which means that zerocopy could break you even in semver-compatible releases. Looks like you actually want [custom validators](https://github.com/google/zerocopy/issues/1330), which are unfortunately not yet implemented. Chime in on that issue if you want to let us know that this is a use case you'd like to support.
